components/brave_shields/adblock/rs/src/result.rs

```diff
@@ -0,0 +1,105 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::str::Utf8Error;
+
+use adblock::{blocker::BlockerError, resources::AddResourceError};
+use thiserror::Error;
+
+use crate::engine::Engine;
+use crate::ffi::*;
+
+#[derive(Debug, Error)]
+pub(crate) enum InternalError {
+  #[error("json error: {0}")]
+  Json(#[from] serde_json::Error),
+  #[error("utf-8 encoding error: {0}")]
+  Utf8(#[from] Utf8Error),
+  #[error("{0}")]
+  Blocker(#[from] BlockerError),
+  #[error("{0}")]
+  AddResource(#[from] AddResourceError),
+}
+
+impl From<&InternalError> for ResultKind {
+  fn from(error: &InternalError) -> Self {
+    match error {
+      InternalError::Json(_) => Self::JsonError,
+      InternalError::Utf8(_) => Self::Utf8Error,
+      InternalError::Blocker(_) | InternalError::AddResource(_) => Self::AdblockError,
+    }
+  }
+}
+
+// This is an intermediate "result" type.
+// The typical chain of conversion for "result" types is as follows:
+// Result<T, InternalError> to PreResult<T> to
+// TResult (i.e. StringResult defined in the ffi module).
+struct PreResult<T: Default> {
+  value: T,
+  result_kind: ResultKind,
+  error_message: String,
+}
+
+// Implements the From<PreResult<T>> trait for a given final ffi result type.
+macro_rules! impl_result_from_trait {
+  ($result_type:ty, $value_type:ty) => {
+    impl From<Result<$value_type, InternalError>> for $result_type {
+      fn from(result: Result<$value_type, InternalError>) -> Self {
+        let PreResult {
+          value,
+          result_kind,
+          error_message,
+        } = PreResult::from(result);
+        Self {
+          value,
+          result_kind,
+          error_message,
+        }
+      }
+    }
+  };
+}
+
+impl<T> From<Result<T, InternalError>> for PreResult<T>
+where
+  T: Default,
+{
+  fn from(result: Result<T, InternalError>) -> Self {
+    match result {
+      Ok(value) => Self {
+        value,
+        result_kind: ResultKind::Success,
+        error_message: String::new(),
+      },
+      Err(e) => Self {
+        value: Default::default(),
```

Comment:
would this mean that accidental access of one of the variables would result in undefined behavior?

it seems that `Vec` and `String` (which are the types that would take up the most space on the stack) are represented by three 8-byte values (assuming a 64-bit arch): pointer to the heap, capacity and length. The total stack size of one of these types would be 24-bytes. No heap allocations occur if the `Vec` or `String` is empty (the pointer within the type is simply a dangling one). `MaybeUninit<T>` also would not reduce any stack memory usage, since it guarantees that the alignment/size is the same as `T`. So it seems the only thing we would save is a couple dozen `mov` instructions to set the three internal values to zero. IMHO, I am unsure if the benefits outweigh the trade-offs of unsafe code. curious to hear your thoughts

Reply:
> would this mean that accidental access of one of the variables would result in undefined behavior?

Correct. Considering that we have this logic to handle the cxx-compatible result types centralized in one place, the amount of unsafe code and thus number of additional invariants to worry about upholding would be relatively small.

> the only thing we would save is a couple dozen mov instructions to set the three internal values to zero

Right, was thinking less in terms of memory and more in terms of instruction count. This mostly applies to the `Error` case - as in, don't initialize an entire `Engine` instance if it's not actually going to be used for anything.

> IMHO, I am unsure if the benefits outweigh the trade-offs of unsafe code.

In retrospect you're probably right. I think I've been doing too much embedded Rust recently; `unsafe` just doesn't have the same weight for me as it used to :sweat_smile: